### ISSUE TITLE
logsoftmax has numerical issue

### ISSUE THREAD

**Issue by zhongwen** [google/jax#985]
Inside stax.py, `logsoftmax` doesn't use the trick of `softmax` where you can do x - x.max to make it more stable.

```
def logsoftmax(x, axis=-1):
  """Apply log softmax to an array of logits, log-normalizing along an axis."""
  return x - logsumexp(x, axis, keepdims=True)
```

```
def softmax(x, axis=-1):
  """Apply softmax to an array of logits, exponentiating and normalizing along an axis."""
  unnormalized = np.exp(x - x.max(axis, keepdims=True))
  return unnormalized / unnormalized.sum(axis, keepdims=True)
```

**Comment by zhongwen** [comment]
still has numerical issue with my proposed implementation.

Will investigate more and give minimal repro here.

**Comment by mattjj** [comment]
[The `logsumexp` function does that trick internally](https://github.com/google/jax/blob/cdb988d2c4d904b29e6e07e91b01cd739d418219/jax/scipy/special.py#L83), which I think should be enough for numerical stability (since that's the only place things are getting exponentiated). Given that, do you think doing it outside as well (as in #986) is redundant?

**Comment by zhongwen** [comment]
Ah, I didn't know that logsumexp has that internally!

I'm having a numerical issue in an experiment, thought it's caused by the logsoftmax. Will investigate other parts then.
